### PR TITLE
Callbacks

### PR DESCRIPTION
**Pull request by pgurenko**
Changed your model a bit to keep the info in arrays

### CODE REVIEW

File: pawnpy/__init__.py [hunk 1]
```diff
@@ -167,82 +181,107 @@ class AMX():
...
-        if numNatives.value != 0:
-            name = create_string_buffer(sNAMEMAX + 1)
-            result = lib.amx_GetNative(byref(self._amx), 0, name, None)
-            if result != AMX_ERR_NONE:
-                raise Exception('amx_GetNative failed with code %d' % result)
+        self._natives = None
+        self.__init_natives()
 
-            print(name.value)
+        self._callbackptr = AMX_CALLBACK(self._callback)
```
Comment (efremovd) [inline pawnpy/__init__.py]: Why do you need it?

File: pawnpy/__init__.py [hunk 2]
```diff
@@ -167,82 +181,107 @@ class AMX():
...
 
-        for i in range(0, num_pub_func.value):
+        for i in range(num_publics.value):
             name = create_string_buffer(sNAMEMAX + 1)
 
             err_code = lib.amx_GetPublic(byref(self._amx), i, name, None)
             if AMX_ERR_NONE != err_code:
                 raise RuntimeError(
-                    "Can't read name of %d public function, code %d" % (i, err_code))
+                    'amx_NumPublics failed with code %d' % result)
```
Comment (efremovd) [inline pawnpy/__init__.py]: Should be amx_GetPublic?

File: pawnpy/__init__.py (hunk 1 above)
Comment (pgurenko) [reply pawnpy/__init__.py]: To save the wrapper - otherwise it will be collected

Comment (efremovd) [reply pawnpy/__init__.py]: No, I meant why you setting callback at all? I I remember correctly that's like advanced topic and we don't need it unless we want to change behavior

File: pawnpy/__init__.py [hunk 3]
```diff
@@ -167,82 +181,107 @@ class AMX():
...
+
+            name = codecs.decode(name.value)
+            self._publics.append(name)
+
+            print('Public function %s' % name)
```
Comment (efremovd) [inline pawnpy/__init__.py]: Do we really need to print this? May be has some util function to print info about AMX?

File: pawnpy/__init__.py [hunk 4]
```diff
@@ -167,82 +181,107 @@ class AMX():
...
-
         for arg in reversed(args):
             # push them into stack
             lib.amx_Push(self._amx, arg)
 
+        index = c_int(func_id)
         ret_val = c_int()
         err_code = lib.amx_Exec(byref(self._amx), byref(ret_val), index)
         if AMX_ERR_NONE != err_code:
+            func_name = self._publics[func_id] if func_id >= 1 else 'main'
```
Comment (efremovd) [inline pawnpy/__init__.py]: Should not be checking against 0?

File: pawnpy/__init__.py (hunk 1 above)
Comment (pgurenko) [reply pawnpy/__init__.py]: You're right - I should use one or the other - let's stick with direct natives hookup

File: pawnpy/__init__.py (hunk 2 above)
Comment (pgurenko) [reply pawnpy/__init__.py]: Yep. Fixed

File: pawnpy/__init__.py (hunk 4 above)
Comment (pgurenko) [reply pawnpy/__init__.py]: Right. Fixed

File: pawnpy/__init__.py (hunk 3 above)
Comment (pgurenko) [reply pawnpy/__init__.py]: Removed - it's debug

**Comment by pgurenko** [comment]
OK, leaving it until tomorrow - looks like Python 3.4.3 behaves differently

File: pawnpy/__init__.py [hunk 5]
```diff
@@ -167,82 +181,96 @@ class AMX():
...
+            if err_code != AMX_ERR_NONE:
+                raise RuntimeError(
+                    'amx_GetNative failed with code %d' % err_code)
+
+            name = codecs.decode(name.value)
+
+            self._natives[i].name = name.encode('utf-8')
+            self._natives[i].func = AMX_NATIVE(self._callback)
+
+        lib.amx_Register(self._amx, self._natives, -1)
```
[the remark this answers is not in the document]
Comment (pgurenko) [reply pawnpy/__init__.py]: Not that simple, tried and still failing. Python 3.5.2 still works...

**Comment by pgurenko** [comment]
Pawn is casting uint32 to ptr in default callback so for now the best way is have our own callback. Not sure how it will work with passing pointers to strings/arrays or whether it will work in this case at all (we might do x86-only build)

**Comment by efremovd** [comment]
so you made it through one callback? overriding default behavior?

**Comment by pgurenko** [comment]
Yep. That's what they actually do, too - setting their own default callback if it is NULL. And in this callback - finding function and calling it, unfortunately casting int to ptr in process

**Pull request merged**